File: #0116_PopulatingNextRightPointersInEachNode/connect.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        head = root
        while root and root.left:
            temp = root.left
            while root:
                root.left.next = root.right
                root.right.next = root.next and root.next.left
                root = root.next
            root = temp
        return head

File: #0116_PopulatingNextRightPointersInEachNode/connect.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public void connect_(Node root){
        if(root == null || root.left == null)
            return;
        root.left.next = root.right;
        if(root.next != null)
            root.right.next = root.next.left;
        connect_(root.left);
        connect_(root.right);
    }
    public Node connect(Node root) {
        connect_(root);
        return root;
    }
}
